### DISCUSSION TITLE
Truck defaults and how to surface warnings instead of failing route

### DISCUSSION THREAD

**Discussion by corypisano** [valhalla/valhalla#4772] [Q&A]
I've been triaging a number of our truck route requests that end up as `442 | No path could be found for input`. 

For an example below, default `truck` options doesn't return a route but `auto` does.
https://valhalla.openstreetmap.de/directions?profile=truck&wps=-64.17509078979494%2C45.800682758808136%2C-63.610925674438484%2C44.70884996592728
![Screen Shot 2024-06-19 at 3 02 02 PM](https://github.com/valhalla/valhalla/assets/1425222/a365dfac-3669-4027-85a5-edf76d682a2d)

Figured out Highway 102 (https://www.openstreetmap.org/way/539225404) has a `maxheight` tagged at 4.1m.  Truck options defaults height to 4.11m if not provided so it just fails. Retrying the request with a height option of 4.1m returns the expected route.
![Screen Shot 2024-06-19 at 3 02 45 PM](https://github.com/valhalla/valhalla/assets/1425222/7dfb4a0d-17e5-44bb-8f85-c946601a259e)

Ideally what I'd like to be able to do is check for truck access (often with truck height/weight/etc unknown) and if there's restrictions that cause the route request to fail entirely, have these surfaced as `warnings`

As of today this would require 3 Valhalla calls is that correct? 
- First `/route` with default truck options
- If no route found, retry `/route` with much less restrictive truck costing options, or perhaps with `ignore_restrictions`/`ignore_access`/etc
- If that returns a route, make a final request to `/trace_attributes` with that polyline and compare against the default truck params to determine where there were restrictions to bubble up as "warnings"

Happy to hear any other suggestions or something I might be missing, thanks!

Some related discussion - https://github.com/valhalla/valhalla/discussions/3767

**Comment by chrstnbwnkl** [comment]
As far as I know an edge's access restrictions can currently not be requested via `/trace_attributes` (though it should be easy to implement). If it were part of the map matching response, you could try and save yourself step 2 and directly call `/trace_attributes` with the relaxed parameters and the same locations as in step 1.

**Reply by chrstnbwnkl** [reply]
Actually, seems like there's already a restriction member in the TripLeg proto https://github.com/valhalla/valhalla/blob/16fecc14df63b6bbb0f53f37c6b31e44589d44cc/proto/trip.proto#L118  in your case it would just require the restriction value to be thrown in there as well plus it would still need to be added to the serializer

**Reply by corypisano** [reply]
ah yes thanks just confirmed `trace_attributes` isn't including `maxheight`, etc. 

For the 2nd point on TripLeg, you mean this would enable getting the restrictions populated in the `/route` response directly? That would be perfect yeah, and then don't need a trace_attributes call

**Reply by corypisano** [reply]
(first time squinting at this path) would this need to also be changed to `repeated`? initially looks like only a single restriction is expected, but for example an edge might have a maxheight and a maxweight tagged, etc

**Reply by chrstnbwnkl** [reply]
I can't tell for sure, but it seems that `restriction` is not repeated because someone wanted to simply include a `has_restriction` in the route summary (similar to `has_ferry` and the likes) since `EnhancedTripPath` has a `has_restriction` flag that is unused (so it never made it to the `/route` response :smile: ).
Anyway, it being present on the Tripleg Edge would enable getting restrictions in both route and trace_attributes responses, it's just a matter of whether it makes sense to have it in the route response. If I understand your use case correctly, you would need to inquire this information in a second request anyway, so calling `trace_attributes` should be fine?

**Reply by chrstnbwnkl** [reply]
And yeah, in that case it would need to be a repeated field
